pkg/controller/networkpolicyrecommendation/controller.go

```diff
@@ -168,11 +300,420 @@ func (c *NPRecommendationController) syncNPRecommendation(key apimachinerytypes.
+	namespacedName := apimachinerytypes.NamespacedName{
+		Name:      npReco.Name,
+		Namespace: npReco.Namespace,
+	}
+	// Stop periodical job
+	c.stopPeriodicSync(namespacedName)
+	// Delete related SparkApplication CR
+	if npReco.Status.SparkApplication == "" {
+		return c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateFailed,
+				ErrorMsg: "Spark Application should be started before updating results",
+			},
+		)
+	}
+
+	if npReco.Status.RecommendedNP == nil {
+		err := deleteSparkApplicationIfExists(c.kubeClient, npReco.Namespace, npReco.Status.SparkApplication)
+		if err != nil {
+			return fmt.Errorf("fail to delete Spark Application: %s, error: %v", npReco.Status.SparkApplication, err)
+		}
+		var recommendedNetworkPolicy *crdv1alpha1.RecommendedNetworkPolicy
+		// Check if RecommendedNetworkPolicy CR is created
+		rnps, err := c.recommendedNPInformer.GetIndexer().ByIndex(idIndex, npReco.Status.SparkApplication)
+		if err == nil && len(rnps) > 0 {
+			// Use the existing RecommendedNetworkPolicy CR
+			recommendedNetworkPolicy = rnps[0].(*crdv1alpha1.RecommendedNetworkPolicy)
+			if len(rnps) > 1 {
+				klog.V(4).InfoS("More than 1 RecommendedNetworkPolicy", "id", npReco.Status.SparkApplication)
+			}
+		} else {
+			if err != nil {
+				klog.V(4).InfoS("Failed to find RecommendedNetworkPolicy", "id", npReco.Status.SparkApplication)
+			}
+			// Get result from database
+			recommendedNetworkPolicy, err = getPolicyRecommendationResult(c.kubeClient, npReco.Status.SparkApplication, npReco.Namespace)
+			if err != nil {
+				return err
+			}
+			// Create RecommendedNetworkPolicy CR
+			recommendedNetworkPolicy, err = c.crdClient.CrdV1alpha1().RecommendedNetworkPolicies(npReco.Namespace).Create(context.TODO(), recommendedNetworkPolicy, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+			klog.V(2).InfoS("Created RecommendedNetworkPolicy", "RecommendedNetworkPolicy", recommendedNetworkPolicy.Name, "NetworkRecommendationPolicy", npReco.Name)
+		}
+		return c.updateNPRecommendationStatus(npReco, crdv1alpha1.NetworkPolicyRecommendationStatus{
+			State:         crdv1alpha1.NPRecommendationStateCompleted,
+			RecommendedNP: recommendedNetworkPolicy,
+			EndTime:       metav1.NewTime(time.Now()),
+		})
+	}
 	return nil
 }
 
+func (c *NPRecommendationController) updateProgress(npReco *crdv1alpha1.NetworkPolicyRecommendation) error {
+	// Check the status before checking the progress in case the job is failed or completed
+	state, err := c.checkSparkApplicationStatus(npReco)
+	if err != nil {
+		return err
+	}
+	if state != crdv1alpha1.NPRecommendationStateRunning {
+		return nil
+	}
+	service := fmt.Sprintf("pr-%s-ui-svc", npReco.Status.SparkApplication)
+	serviceIP, servicePort, err := getServiceAddr(c.kubeClient, service, npReco.Namespace, sparkPortName)
```

Comment:
feels like it's a bit wasteful given the service addr won't change, yet multiple api calls to k8s apiserver is done each time we do periodical sync. Also given that manager is now running in k8s cluster (in hence have access to coreDNS), can we rather access the spark svc via DNS name?

Reply:
Updated, thanks!